### ISSUE TITLE
Add attributes to control Swizzle and SplitK

### ISSUE THREAD

**Issue by powderluv** [iree-org/iree#12405]
### Request description

SHARK-Runtime has support for controlling Swizzle and SplitK 
https://github.com/openxla/iree/commit/5a6c22a4ae81ea1974a0de2167d6d1bbd39fefc8
https://github.com/openxla/iree/commit/2a36f3227d6cbaad85286c670cfae274e81b26a4

We would like to have these two upstreamed cleanly.

**Comment by ThomasRaoux** [comment]
@MaheshRavishankar I think most of the concerns about splitK attributes were coming from you. I see that you added some  attributes to control fusion here:
https://github.com/openxla/iree/blame/main/compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp#L89

Are you fine with me adding similar attributes to control splitK?

**Comment by MaheshRavishankar** [comment]
> @MaheshRavishankar I think most of the concerns about splitK attributes were coming from you. I see that you added some attributes to control fusion here: [`main`/compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp#L89 (blame)](https://github.com/openxla/iree/blame/main/compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp#L89)
> 
> Are you fine with me adding similar attributes to control splitK?

Those attributes are pass internal and not meant to be set externally. So they are not the same thing....

**Comment by ThomasRaoux** [comment]
> > @MaheshRavishankar I think most of the concerns about splitK attributes were coming from you. I see that you added some attributes to control fusion here: [`main`/compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp#L89 (blame)](https://github.com/openxla/iree/blame/main/compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp#L89)
> > Are you fine with me adding similar attributes to control splitK?
> 
> Those attributes are pass internal and not meant to be set externally. So they are not the same thing....

ok, so we still don't want to allow splitK kind of attributes?

**Comment by MaheshRavishankar** [comment]
> > > @MaheshRavishankar I think most of the concerns about splitK attributes were coming from you. I see that you added some attributes to control fusion here: [`main`/compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp#L89 (blame)](https://github.com/openxla/iree/blame/main/compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp#L89)
> > > Are you fine with me adding similar attributes to control splitK?
> > 
> > 
> > Those attributes are pass internal and not meant to be set externally. So they are not the same thing....
> 
> ok, so we still don't want to allow splitK kind of attributes?

I am not sure I can answer that question. I understand why you need it... but I really am not for adding more attributes that control things at Flow level this way (it really points to layering issue with split-k). This seems ad-hoc and I don't want this to be considered as a supported feature....

**Comment by ThomasRaoux** [comment]
> Flow level this way (it really points to layering issue with split-k)

The layering issue is that splitK should be in pre-processing?

**Comment by MaheshRavishankar** [comment]
> > Flow level this way (it really points to layering issue with split-k)
> 
> The layering issue is that splitK should be in pre-processing?

Even if you do that it might not help you cause I think you wanted it to run after elementwise fusion. That would be great indeed, then you dont need an attribute at all. Its a pass option you pass as a string.

**Comment by ThomasRaoux** [comment]
> Even if you do that it might not help you cause I think you wanted it to run after elementwise fusion. That would be great indeed, then you dont need an attribute at all. Its a pass option you pass as a string.

I don't need it to run after elementwise fusion however I do need it to run after the pass collapsing reduction dimension. I guess this could also be ran in pre-processing.

That being said we would still need an attribute as not all ops would want the same splitK parameter so it still needs to be a per op attribute. Is having an attribute acceptable to you if the pass moves to pre-processing?

**Comment by MaheshRavishankar** [comment]
> > Even if you do that it might not help you cause I think you wanted it to run after elementwise fusion. That would be great indeed, then you dont need an attribute at all. Its a pass option you pass as a string.
> 
> I don't need it to run after elementwise fusion however I do need it to run after the pass collapsing reduction dimension. I guess this could also be ran in pre-processing.
> 
> That being said we would still need an attribute as not all ops would want the same splitK parameter so it still needs to be a per op attribute. Is having an attribute acceptable to you if the pass moves to pre-processing?

Is that an option really. This means that all downstream "power" users will have to add it to their pre-processing pass-pipeline, correct? (btw, I might not get to it soon, but renaming preprocessing to specialization would be a great things to do!)

**Comment by ThomasRaoux** [comment]
> This means that all downstream "power" users will have to add it to their pre-processing pass-pipeline, correct?

Right, this is equivalent to the tuning attributes. Ideally we should also have a default heuristic that allows power users to get the transformation to happen without having to do anything but that requires extra infrastructure to allow supporting platform influenced heuristics.

**Comment by MaheshRavishankar** [comment]
> > This means that all downstream "power" users will have to add it to their pre-processing pass-pipeline, correct?
> 
> Right, this is equivalent to the tuning attributes. Ideally we should also have a default heuristic that allows power users to get the transformation to happen without having to do anything but that requires extra infrastructure to allow supporting platform influenced heuristics.

What is the attribute to control swizzling?

**Comment by ThomasRaoux** [comment]
> > > This means that all downstream "power" users will have to add it to their pre-processing pass-pipeline, correct?
> > 
> > 
> > Right, this is equivalent to the tuning attributes. Ideally we should also have a default heuristic that allows power users to get the transformation to happen without having to do anything but that requires extra infrastructure to allow supporting platform influenced heuristics.
> 
> What is the attribute to control swizzling?

There isn't one right now but since this is a codegen knob we want to tune for matmul codegen we would need to add one (similar to software pipelining depth)

**Comment by MaheshRavishankar** [comment]
> > This means that all downstream "power" users will have to add it to their pre-processing pass-pipeline, correct?
> 
> Right, this is equivalent to the tuning attributes. Ideally we should also have a default heuristic that allows power users to get the transformation to happen without having to do anything but that requires extra infrastructure to allow supporting platform influenced heuristics.

I am still on the fence about having such attributes in Flow. I think it ends up setting a wrong example (already every one complains about the current attribute in codegen itself, so adding one similar to that at the Flow level is hard to justify). One option might be to put this in preprocessing/specialization. Then its a something that you add for specialization. Is clear demarcation to suggest that the attribute is a WAR... Dont know if that works for Nod.

**Comment by benvanik** [comment]
yeah, no flow level attributes like this - fusion control is applicable because they're generic language-style features like `inline` or `__device__` that apply to how a program is formed, while something like splitk is problem specific. You'd either be changing something before flow during preprocessing in a source dialect and it can be there (mhlo/linalg/etc) or changing something in codegen during translation and it can be there. We respect dialect attrs in most places - if you're just wanting to squirrel across extra information you can add `my.whatever = #whatever` to any op and it'll end up in the backend. If we are dropping them that's a bug we can fix.

**Comment by benvanik** [comment]
(if there's something specific during dispatch region formation you need to control then we could try to distill what you're actually wanting to control into a general form - but I think the referenced example looks like a codegen thing)

**Comment by ThomasRaoux** [comment]
> I am still on the fence about having such attributes in Flow. I think it ends up setting a wrong example (already every one complains about the current attribute in codegen itself, so adding one similar to that at the Flow level is hard to justify). One option might be to put this in preprocessing/specialization. Then its a something that you add for specialization. Is clear demarcation to suggest that the attribute is a WAR... Dont know if that works for Nod.

I agree attributes are painful and are likely to break in the future (for instance if we move to generalizing matmul ops or if we want to tune some linalg.generic ops). However I don't think we have a replacement for those?

> yeah, no flow level attributes like this - fusion control is applicable because they're generic language-style features like `inline` or `__device__` that apply to how a program is formed, while something like splitk is problem specific. You'd either be changing something before flow during preprocessing in a source dialect and it can be there (mhlo/linalg/etc) or changing something in codegen during translation and it can be there. We respect dialect attrs in most places - if you're just wanting to squirrel across extra information you can add `my.whatever = #whatever` to any op and it'll end up in the backend. If we are dropping them that's a bug we can fix.

The problem is that splitK needs to happen before codegen. (needs to create 2 dispatches). Are you saying that if this pass was not in flow then using attributes to control is acceptable?

**Comment by ThomasRaoux** [comment]
> (if there's something specific during dispatch region formation you need to control then we could try to distill what you're actually wanting to control into a general form - but I think the referenced example looks like a codegen thing)

It might confusing as there are two cases in this bug but the splitK part needs to happen before dispatch region formation.

**Comment by benvanik** [comment]
that pass should be in the preprocessing layer and not in flow - it's just doing linalg stuff

**Comment by ThomasRaoux** [comment]
> that pass should be in the preprocessing layer and not in flow - it's just doing linalg stuff

yep I agree. Is adding attribute there okay with everybody?

**Comment by benvanik** [comment]
I'd have to see what's being proposed, but generally anything that happens before flow is fair game for whatever you want to do (hand-authoring entire dispatches, partially forming dispatch regions you know you want, etc)

**Comment by MaheshRavishankar** [comment]
> > that pass should be in the preprocessing layer and not in flow - it's just doing linalg stuff
> 
> yep I agree. Is adding attribute there okay with everybody?

Ok in the sense that its still a WAR... I'd really hope we can fix the split-k layering by using reduction tiling, but will require work that is not scoped out. So its fine by me.

**Comment by ThomasRaoux** [comment]
> > > that pass should be in the preprocessing layer and not in flow - it's just doing linalg stuff
> > 
> > 
> > yep I agree. Is adding attribute there okay with everybody?
> 
> Ok in the sense that its still a WAR... I'd really hope we can fix the split-k layering by using reduction tiling, but will require work that is not scoped out. So its fine by me.

How would using reduction tiling help this case? We still need to control the transformation even if using the reduction tiling transform.


> I'd have to see what's being proposed, but generally anything that happens before flow is fair game for whatever you want to do (hand-authoring entire dispatches, partially forming dispatch regions you know you want, etc)

it would look like this:

```
%10 = linalg.matmul  attrs =  {split_k_ratio = 4}
```

**Comment by benvanik** [comment]
that seems fine - you'd probably want to make it a dialect attribute so it's preserved properly (`iree_codegen.split_k_ratio = 4 : index` or whatever)

**Comment by allieculp** [comment]
@ThomasRaoux Can you update the status here? Is this in progress?

**Comment by ThomasRaoux** [comment]
This hasn't started yet. I'm not sure of the priority at this time, we have other discussions on split-k and it sounds like there are some basic disconnect on the vision of how it should be done. I would like this to be resolved before spending time on that. 
maybe @mattwalsh can help getting this sorted out?

**Comment by allieculp** [comment]
Thanks Thomas, I may pull this into our next sprint meeting for continued priority discussion!

**Comment by mattwalsh** [comment]
Yeah I would say given we are covering this space now there's no danger of
forgetting to do the work this older ticket called out and it's ok to
delete.

**Comment by allieculp** [comment]
@manishucsd Can you take a look at this?

**Comment by allieculp** [comment]
@manishucsd Any update on this one?

**Comment by manishucsd** [comment]
The split-k part of this being discussed here #13115.
I am not sure what Swizzle means here? Shared Memory swizzle or Threadblock L2 swizzle? 
- Shared memory swizzle should also be enabled to avoid bank-conflict. 
- Threadblock l2 swizzle can be controlled and that's what the tile seems to be?

**Comment by allieculp** [comment]
Maybe we can close this in favor of #13115 ?

**Comment by MaheshRavishankar** [comment]
Lets close this and have a single bug for all split-k related issues (I cant track what bug tracks what specific issue)

**Comment by powderluv** [comment]
This issue is to specifically add attribute control so we drop the patches listed in the first comment. I am not sure if that is handled by #13115 but if it is great if not lets leave this open.

**Comment by ThomasRaoux** [comment]
Yes this is orthogonal to https://github.com/openxla/iree/issues/13115, re-opening the issue

**Comment by MaheshRavishankar** [comment]
I think there are too many bugs related to these aspects. We need to consolidate these. I for one am lost in all the bugs related to this and their variants.

**Comment by allieculp** [comment]
@manishucsd Should we leave this one open? Is there additional split-k work here?

**Comment by allieculp** [comment]
@qedawkins @powderluv Is this issue time-sensitive for Nod? We are working on sprint planning and looking for prioritization.

**Comment by powderluv** [comment]
not time-sensitive. We have a local patch.
